File: android/app/src/main/kotlin/com/amazonaws/services/chime/flutterdemo/MethodChannelCoordinator.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 */
 
package com.amazonaws.services.chime.flutterdemo

import com.amazonaws.services.chime.sdk.meetings.device.MediaDevice
import com.amazonaws.services.chime.sdk.meetings.session.DefaultMeetingSession
import com.amazonaws.services.chime.sdk.meetings.session.MediaPlacement
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionConfiguration
import com.amazonaws.services.chime.sdk.meetings.session.CreateMeetingResponse
import com.amazonaws.services.chime.sdk.meetings.session.Meeting
import com.amazonaws.services.chime.sdk.meetings.session.CreateAttendeeResponse
import com.amazonaws.services.chime.sdk.meetings.session.Attendee
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger
import com.amazonaws.services.chime.flutterdemo.MethodCall as MethodCallFlutter
import android.app.Activity
import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import androidx.appcompat.app.AppCompatActivity
import io.flutter.plugin.common.MethodChannel

class MethodChannelCoordinator(binaryMessenger: BinaryMessenger, activity: Activity) :
    AppCompatActivity() {
    val methodChannel: MethodChannel
    val context: Context
    var permissionsManager: PermissionManager = PermissionManager(activity)

    init {
        methodChannel =
            MethodChannel(binaryMessenger, "com.amazonaws.services.chime.flutterDemo.methodChannel")
        context = activity.applicationContext
    }

    private val NULL_MEETING_SESSION_RESPONSE: MethodChannelResult =
        MethodChannelResult(false, Response.meeting_session_is_null.msg)

    fun setupMethodChannel() {
        methodChannel.setMethodCallHandler { call, result ->
            val callResult: MethodChannelResult
            when (call.method) {
                MethodCallFlutter.manageAudioPermissions.call -> {
                    permissionsManager.manageAudioPermissions(result)
                    return@setMethodCallHandler
                }
                MethodCallFlutter.manageVideoPermissions.call -> {
                    permissionsManager.manageVideoPermissions(result)
                    return@setMethodCallHandler
                }
                MethodCallFlutter.join.call -> {
                    callResult = join(call)
                }
                MethodCallFlutter.stop.call -> {
                    callResult = stop()
                }
                MethodCallFlutter.mute.call -> {
                    callResult = mute()
                }
                MethodCallFlutter.unmute.call -> {
                    callResult = unmute()
                }
                MethodCallFlutter.startLocalVideo.call -> {
                    callResult = startLocalVideo()
                }
                MethodCallFlutter.stopLocalVideo.call -> {
                    callResult = stopLocalVideo()
                }
                MethodCallFlutter.initialAudioSelection.call -> {
                    callResult = initialAudioSelection()
                }
                MethodCallFlutter.listAudioDevices.call -> {
                    callResult = listAudioDevices()
                }
                MethodCallFlutter.updateAudioDevice.call -> {
                    callResult = updateAudioDevice(call)
                }
                else -> callResult = MethodChannelResult(false, Response.method_not_implemented)
            }

            if (callResult.result) {
                result.success(callResult.toFlutterCompatibleType())
            } else {
                result.error(
                    "Failed",
                    "MethodChannelHandler failed",
                    callResult.toFlutterCompatibleType()
                )
            }
        }
    }

    fun callFlutterMethod(method: MethodCallFlutter, args: Any?) {
        methodChannel.invokeMethod(method.call, args)
    }

    fun join(call: MethodCall): MethodChannelResult {
        if (call.arguments == null) {
            return MethodChannelResult(false, Response.incorrect_join_response_params.msg)
        }
        val meetingId: String? = call.argument("MeetingId")
        val externalMeetingId: String? = call.argument("ExternalMeetingId")
        val mediaRegion: String? = call.argument("MediaRegion")
        val audioHostUrl: String? = call.argument("AudioHostUrl")
        val audioFallbackUrl: String? = call.argument("AudioFallbackUrl")
        val signalingUrl: String? = call.argument("SignalingUrl")
        val turnControlUrl: String? = call.argument("TurnControlUrl")
        val externalUserId: String? = call.argument("ExternalUserId")
        val attendeeId: String? = call.argument("AttendeeId")
        val joinToken: String? = call.argument("JoinToken")

        if (meetingId == null ||
            mediaRegion == null ||
            audioHostUrl == null ||
            externalMeetingId == null ||
            audioFallbackUrl == null ||
            signalingUrl == null ||
            turnControlUrl == null ||
            externalUserId == null ||
            attendeeId == null ||
            joinToken == null
        ) {
            return MethodChannelResult(false, Response.incorrect_join_response_params.msg)
        }

        val createMeetingResponse = CreateMeetingResponse(
            Meeting(
                externalMeetingId,
                MediaPlacement(audioFallbackUrl, audioHostUrl, signalingUrl, turnControlUrl),
                mediaRegion,
                meetingId
            )
        )
        val createAttendeeResponse =
            CreateAttendeeResponse(Attendee(attendeeId, externalUserId, joinToken))
        val meetingSessionConfiguration =
            MeetingSessionConfiguration(createMeetingResponse, createAttendeeResponse)

        val meetingSession =
            DefaultMeetingSession(meetingSessionConfiguration, ConsoleLogger(), context)

        MeetingSessionManager.meetingSession = meetingSession
        return MeetingSessionManager.startMeeting(
            RealtimeObserver(this),
            VideoTileObserver(this),
            AudioVideoObserver(this)
        )
    }

    fun stop(): MethodChannelResult {
        return MeetingSessionManager.stop()
    }

    fun mute(): MethodChannelResult {
        val muted = MeetingSessionManager.meetingSession?.audioVideo?.realtimeLocalMute()
            ?: return NULL_MEETING_SESSION_RESPONSE
        return if (muted) MethodChannelResult(
            true,
            Response.mute_successful.msg
        ) else MethodChannelResult(false, Response.mute_failed.msg)
    }

    fun unmute(): MethodChannelResult {
        val unmuted = MeetingSessionManager.meetingSession?.audioVideo?.realtimeLocalUnmute()
            ?: return NULL_MEETING_SESSION_RESPONSE
        return if (unmuted) MethodChannelResult(
            true,
            Response.unmute_successful.msg
        ) else MethodChannelResult(false, Response.unmute_failed.msg)
    }

    fun startLocalVideo(): MethodChannelResult {
        MeetingSessionManager.meetingSession?.audioVideo?.startLocalVideo()
            ?: return NULL_MEETING_SESSION_RESPONSE
        return MethodChannelResult(true, Response.local_video_on_success.msg)
    }

    fun stopLocalVideo(): MethodChannelResult {
        MeetingSessionManager.meetingSession?.audioVideo?.stopLocalVideo()
            ?: return NULL_MEETING_SESSION_RESPONSE
        return MethodChannelResult(true, Response.local_video_on_success.msg)
    }

    fun initialAudioSelection(): MethodChannelResult {
        val device =
            MeetingSessionManager.meetingSession?.audioVideo?.getActiveAudioDevice()
                ?: return NULL_MEETING_SESSION_RESPONSE
        return MethodChannelResult(true, device.label)
    }

    fun listAudioDevices(): MethodChannelResult {
        val audioDevices = MeetingSessionManager.meetingSession?.audioVideo?.listAudioDevices()
            ?: return NULL_MEETING_SESSION_RESPONSE
        val transform: (MediaDevice) -> String = { it.label }
        return MethodChannelResult(true, audioDevices.map(transform))
    }

    fun updateAudioDevice(call: MethodCall): MethodChannelResult {
        val device =
            call.arguments ?: return MethodChannelResult(false, Response.null_audio_device.msg)

        val audioDevices = MeetingSessionManager.meetingSession?.audioVideo?.listAudioDevices()
            ?: return NULL_MEETING_SESSION_RESPONSE

        for (dev in audioDevices) {
            if (device == dev.label) {
                MeetingSessionManager.meetingSession?.audioVideo?.chooseAudioDevice(dev)
                    ?: return MethodChannelResult(false, Response.audio_device_update_failed.msg)
                return MethodChannelResult(true, Response.audio_device_updated.msg)
            }
        }
        return MethodChannelResult(false, Response.audio_device_update_failed.msg)
    }
}

File: ios/Runner/MethodChannelCoordinator.swift
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: MIT-0
 */
 
import AmazonChimeSDK
import AmazonChimeSDKMedia
import AVFoundation
import Flutter
import Foundation
import UIKit

class MethodChannelCoordinator {
    let methodChannel: FlutterMethodChannel
    
    var realtimeObserver: RealtimeObserver?
    
    var audioVideoObserver: AudioVideoObserver?
    
    var videoTileObserver: VideoTileObserver?
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.methodChannel = FlutterMethodChannel(name: "com.amazonaws.services.chime.flutterDemo.methodChannel", binaryMessenger: binaryMessenger)
    }
    
    //
    // ————————————————————————————————— Method Call Setup —————————————————————————————————
    //
    
    func setUpMethodCallHandler() {
        self.methodChannel.setMethodCallHandler { [unowned self]
            (call: FlutterMethodCall, result: @escaping FlutterResult) in
                let callMethod = MethodCall(rawValue: call.method)
                var response: MethodChannelResponse = .init(result: false, arguments: nil)
                switch callMethod {
                case .manageAudioPermissions:
                    response = self.manageAudioPermissions()
                case .manageVideoPermissions:
                    response = self.manageVideoPermissions()
                case .join:
                    response = self.join(call: call)
                case .stop:
                    response = self.stop()
                case .mute:
                    response = self.mute()
                case .unmute:
                    response = self.unmute()
                case .startLocalVideo:
                    response = self.startLocalVideo()
                case .stopLocalVideo:
                    response = self.stopLocalVideo()
                case .initialAudioSelection:
                    response = self.initialAudioSelection()
                case .listAudioDevices:
                    response = self.listAudioDevices()
                case .updateAudioDevice:
                    response = self.updateAudioDevice(call: call)
                default:
                    response = MethodChannelResponse(result: false, arguments: Response.method_not_implemented)
                }
                result(response.toFlutterCompatibleType())
        }
    }
    
    func callFlutterMethod(method: MethodCall, args: Any?) {
        self.methodChannel.invokeMethod(method.rawValue, arguments: args)
    }
    
    //
    // ————————————————————————————————— Method Call Options —————————————————————————————————
    //
    
    func manageAudioPermissions() -> MethodChannelResponse {
        let audioPermission = AVAudioSession.sharedInstance().recordPermission
        switch audioPermission {
        case .undetermined:
            if self.requestAudioPermission() {
                return MethodChannelResponse(result: true, arguments: Response.audio_authorized.rawValue)
            }
            return MethodChannelResponse(result: false, arguments: Response.audio_auth_not_granted.rawValue)
        case .granted:
            return MethodChannelResponse(result: true, arguments: Response.audio_authorized.rawValue)
        case .denied:
            return MethodChannelResponse(result: false, arguments: Response.audio_auth_not_granted.rawValue)
        @unknown default:
            return MethodChannelResponse(result: false, arguments: Response.unknown_audio_authorization_status.rawValue)
        }
    }
    
    func manageVideoPermissions() -> MethodChannelResponse {
        let videoPermission: AVAuthorizationStatus = AVCaptureDevice.authorizationStatus(for: .video)
        switch videoPermission {
        case .notDetermined:
            if self.requestVideoPermission() {
                return MethodChannelResponse(result: true, arguments: Response.video_authorized.rawValue)
            }
            return MethodChannelResponse(result: false, arguments: Response.video_auth_not_granted.rawValue)
        case .authorized:
            return MethodChannelResponse(result: true, arguments: Response.video_authorized.rawValue)
        case .denied:
            return MethodChannelResponse(result: false, arguments: Response.video_auth_not_granted.rawValue)
        case .restricted:
            return MethodChannelResponse(result: false, arguments: Response.video_restricted.rawValue)
        @unknown default:
            return MethodChannelResponse(result: false, arguments: Response.unknown_video_authorization_status.rawValue)
        }
    }
    
    func join(call: FlutterMethodCall) -> MethodChannelResponse {
        guard let json = call.arguments as? [String: String] else {
            return MethodChannelResponse(result: false, arguments: Response.create_meeting_failed)
        }
        
        // TODO: zmauricv: add a Json Decoder
        guard let meetingId = json["MeetingId"], let externalMeetingId = json["ExternalMeetingId"], let mediaRegion = json["MediaRegion"], let audioHostUrl = json["AudioHostUrl"], let audioFallbackUrl = json["AudioFallbackUrl"], let signalingUrl = json["SignalingUrl"], let turnControlUrl = json["TurnControlUrl"], let externalUserId = json["ExternalUserId"], let attendeeId = json["AttendeeId"], let joinToken = json["JoinToken"]
        else {
            return MethodChannelResponse(result: false, arguments: Response.incorrect_join_response_params.rawValue)
        }
        
        let meetingResponse = CreateMeetingResponse(meeting: Meeting(externalMeetingId: externalMeetingId, mediaPlacement: MediaPlacement(audioFallbackUrl: audioFallbackUrl, audioHostUrl: audioHostUrl, signalingUrl: signalingUrl, turnControlUrl: turnControlUrl), mediaRegion: mediaRegion, meetingId: meetingId))
        
        let attendeeResponse = CreateAttendeeResponse(attendee: Attendee(attendeeId: attendeeId, externalUserId: externalUserId, joinToken: joinToken))
        
        let meetingSessionConfiguration = MeetingSessionConfiguration(createMeetingResponse: meetingResponse, createAttendeeResponse: attendeeResponse)
        
        let logger = ConsoleLogger(name: "MeetingSession Logger", level: LogLevel.DEBUG)
        
        let meetingSession = DefaultMeetingSession(configuration: meetingSessionConfiguration, logger: logger)
        
        self.configureAudioSession()
        
        // Update Singleton Class
        MeetingSession.shared.meetingSession = meetingSession
        
        self.setupAudioVideoFacadeObservers()
        let meetingStartResponse = MeetingSession.shared.startMeetingAudio()
        
        return meetingStartResponse
    }
    
    func stop() -> MethodChannelResponse {
        MeetingSession.shared.meetingSession?.audioVideo.stop()
        MeetingSession.shared.meetingSession = nil
        return MethodChannelResponse(result: true, arguments: Response.meeting_stopped_successfully.rawValue)
    }
    
    func mute() -> MethodChannelResponse {
        let muted = MeetingSession.shared.meetingSession?.audioVideo.realtimeLocalMute() ?? false
        
        if muted {
            return MethodChannelResponse(result: true, arguments: Response.mute_successful.rawValue)
        } else {
            return MethodChannelResponse(result: false, arguments: Response.mute_failed.rawValue)
        }
    }
    
    func unmute() -> MethodChannelResponse {
        let unmuted = MeetingSession.shared.meetingSession?.audioVideo.realtimeLocalUnmute() ?? false
        
        if unmuted {
            return MethodChannelResponse(result: true, arguments: Response.unmute_successful.rawValue)
        } else {
            return MethodChannelResponse(result: false, arguments: Response.unmute_successful.rawValue)
        }
    }
    
    func startLocalVideo() -> MethodChannelResponse {
        do {
            try MeetingSession.shared.meetingSession?.audioVideo.startLocalVideo()
            return MethodChannelResponse(result: true, arguments: Response.local_video_on_success.rawValue)
        } catch {
            MeetingSession.shared.meetingSession?.logger.error(msg: "Error configuring AVAudioSession: \(error.localizedDescription)")
            return MethodChannelResponse(result: false, arguments: Response.local_video_on_failed.rawValue)
        }
    }
    
    func stopLocalVideo() -> MethodChannelResponse {
        MeetingSession.shared.meetingSession?.audioVideo.stopLocalVideo()
        return MethodChannelResponse(result: true, arguments: Response.local_video_off_success.rawValue)
    }
    
    func initialAudioSelection() -> MethodChannelResponse {
        if let initialAudioDevice = MeetingSession.shared.meetingSession?.audioVideo.getActiveAudioDevice() {
            return MethodChannelResponse(result: true, arguments: initialAudioDevice.label)
        }
        return MethodChannelResponse(result: false, arguments: Response.failed_to_get_initial_audio_device.rawValue)
    }
    
    func listAudioDevices() -> MethodChannelResponse {
        guard let audioDevices = MeetingSession.shared.meetingSession?.audioVideo.listAudioDevices() else {
            return MethodChannelResponse(result: false, arguments: Response.failed_to_list_audio_devices.rawValue)
        }
        
        return MethodChannelResponse(result: true, arguments: audioDevices.map { $0.label })
    }
    
    func updateAudioDevice(call: FlutterMethodCall) -> MethodChannelResponse {
        guard let device = call.arguments as? String else {
            return MethodChannelResponse(result: false, arguments: Response.audio_device_update_failed.rawValue)
        }
        
        guard let audioDevices = MeetingSession.shared.meetingSession?.audioVideo.listAudioDevices() else {
            MeetingSession.shared.meetingSession?.logger.error(msg: Response.failed_to_list_audio_devices.rawValue)
            return MethodChannelResponse(result: false, arguments: Response.failed_to_list_audio_devices.rawValue)
        }
        
        for dev in audioDevices {
            if device == dev.label {
                MeetingSession.shared.meetingSession?.audioVideo.chooseAudioDevice(mediaDevice: dev)
                return MethodChannelResponse(result: true, arguments: Response.audio_device_updated.rawValue)
            }
        }
        
        return MethodChannelResponse(result: false, arguments: Response.audio_device_update_failed.rawValue)
    }
    
    //
    // ————————————————————————————————— Helper Functions —————————————————————————————————
    //
    
    private func requestAudioPermission() -> Bool {
        var result = false
        
        let group = DispatchGroup()
        group.enter()
        DispatchQueue.global(qos: .default).async {
            AVAudioSession.sharedInstance().requestRecordPermission { granted in
                result = granted
                group.leave()
            }
        }
        group.wait()
        return result
    }
    
    private func requestVideoPermission() -> Bool {
        var result = false
        
        let group = DispatchGroup()
        group.enter()
        DispatchQueue.global(qos: .default).async {
            AVCaptureDevice.requestAccess(for: .video) { granted in
                result = granted
                group.leave()
            }
        }
        group.wait()
        return result
    }
    
    private func setupAudioVideoFacadeObservers() {
        self.realtimeObserver = MyRealtimeObserver(withMethodChannel: self)
        if self.realtimeObserver != nil {
            MeetingSession.shared.meetingSession?.audioVideo.addRealtimeObserver(observer: self.realtimeObserver!)
            MeetingSession.shared.meetingSession?.logger.info(msg: "realtimeObserver set up...")
        }
        
        self.audioVideoObserver = MyAudioVideoObserver(withMethodChannel: self)
        if self.audioVideoObserver != nil {
            MeetingSession.shared.meetingSession?.audioVideo.addAudioVideoObserver(observer: self.audioVideoObserver!)
            MeetingSession.shared.meetingSession?.logger.info(msg: "audioVideoObserver set up...")
        }
        
        self.videoTileObserver = MyVideoTileObserver(withMethodChannel: self)
        if self.videoTileObserver != nil {
            MeetingSession.shared.meetingSession?.audioVideo.addVideoTileObserver(observer: self.videoTileObserver!)
            MeetingSession.shared.meetingSession?.logger.info(msg: "VideoTileObserver set up...")
        }
    }
    
    func stopAudioVideoFacadeObservers() {
        if let rtObserver = self.realtimeObserver {
            MeetingSession.shared.meetingSession?.audioVideo.removeRealtimeObserver(observer: rtObserver)
        }
        
        if let avObserver = self.audioVideoObserver {
            MeetingSession.shared.meetingSession?.audioVideo.removeAudioVideoObserver(observer: avObserver)
        }
        
        if let vtObserver = self.videoTileObserver {
            MeetingSession.shared.meetingSession?.audioVideo.removeVideoTileObserver(observer: vtObserver)
        }
    }
    
    private func configureAudioSession() {
        let audioSession = AVAudioSession.sharedInstance()
        do {
            if audioSession.category != .playAndRecord {
                try audioSession.setCategory(AVAudioSession.Category.playAndRecord,
                                             options: AVAudioSession.CategoryOptions.allowBluetooth)
                try audioSession.setActive(true, options: .notifyOthersOnDeactivation)
            }
            if audioSession.mode != .voiceChat {
                try audioSession.setMode(.voiceChat)
            }
        } catch {
            MeetingSession.shared.meetingSession?.logger.error(msg: "Error configuring AVAudioSession: \(error.localizedDescription)")
        }
    }
}
